dnf5-plugins/system_upgrade_plugin/system_upgrade.cpp

```diff
@@ -0,0 +1,517 @@
+
+void SystemUpgradeSubcommand::set_argument_parser() {
+    auto & ctx = get_context();
+    auto & parser = ctx.get_argument_parser();
+    auto & cmd = *get_argument_parser_command();
+
+    cachedir =
+        dynamic_cast<libdnf5::OptionPath *>(parser.add_init_value(std::make_unique<libdnf5::OptionPath>(datadir)));
+
+    auto * download_dir_arg = parser.add_new_named_arg("downloaddir");
+    download_dir_arg->set_long_name("downloaddir");
+    download_dir_arg->set_description("Redirect download of packages to provided <path>");
+    download_dir_arg->link_value(cachedir);
+    cmd.register_named_arg(download_dir_arg);
+}
+
+void SystemUpgradeSubcommand::configure() {
+    auto & ctx = get_context();
+    const std::filesystem::path installroot{ctx.base.get_config().get_installroot_option().get_value()};
+    magic_symlink = installroot / "system-update";
+}
+
+void SystemUpgradeDownloadCommand::set_argument_parser() {
+    SystemUpgradeSubcommand::set_argument_parser();
+
+    auto & ctx = get_context();
+    auto & parser = ctx.get_argument_parser();
+    auto & cmd = *get_argument_parser_command();
+
+    cmd.set_description("Downloads everything needed to upgrade to a new release");
+
+    no_downgrade =
+        dynamic_cast<libdnf5::OptionBool *>(parser.add_init_value(std::make_unique<libdnf5::OptionBool>(true)));
+
+    auto * no_downgrade_arg = parser.add_new_named_arg("no-downgrade");
+    no_downgrade_arg->set_long_name("no-downgrade");
+    no_downgrade_arg->set_description(
+        "Do not install packages from the new release if they are older than what is currently installed");
+    no_downgrade_arg->link_value(no_downgrade);
+
+    cmd.register_named_arg(no_downgrade_arg);
+}
+
+void SystemUpgradeDownloadCommand::configure() {
+    SystemUpgradeSubcommand::configure();
+
+    auto & ctx = get_context();
+    auto & conf = ctx.base.get_config();
+
+    // Check --releasever
+    const auto & installroot = conf.get_installroot_option().get_value();
+    system_releasever = *libdnf5::Vars::detect_release(ctx.base.get_weak_ptr(), installroot);
+    target_releasever = ctx.base.get_vars()->get_value("releasever");
+
+    if (target_releasever == system_releasever) {
+        throw libdnf5::cli::CommandExitError(1, M_("Need a --releasever greater than the current system version."));
+    }
+
+    ctx.set_load_system_repo(true);
+    ctx.set_load_available_repos(Context::LoadAvailableRepos::ENABLED);
+
+    ctx.base.get_config().get_cachedir_option().set(
+        libdnf5::Option::Priority::PLUGINDEFAULT, get_cachedir()->get_value());
+}
+
+void SystemUpgradeDownloadCommand::run() {
+    auto & ctx = get_context();
+
+    const auto & goal = std::make_unique<libdnf5::Goal>(ctx.base);
+
+    auto & conf = ctx.base.get_config();
+    conf.opt_binds().at("tsflags").new_string(libdnf5::Option::Priority::RUNTIME, "test");
+
+    if (no_downgrade->get_value()) {
+        const auto & settings = libdnf5::GoalJobSettings();
```

Comment:
Isn't the `add_rpm_upgrade` by default adding an empty settings?

Reply:
Yes, thanks